### PR TITLE
8295267: Improve G1 scan to merge ratio calculation

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that fixes the merge-to-scan ratio calculation ("which fraction of the remembered sets creates unique cards") for predictions to make it more accurate and understandable?

Previously the denominator has been the total number of scanned cards; this seems to be a good approximation (trying to make a case for it, it's wrong) if the number of log buffers is low at GC start, but tends to get broken otherwise (i.e. with [JDK-8137022](https://bugs.openjdk.org/browse/JDK-8137022)).

This change fixes this to use the number of cards that were merged in.

This change depends on PR#10656. Please review that one first if not done otherwise.

Testing: tier1, checking that no tests use the "Dirty Cards" string now changed.

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295267](https://bugs.openjdk.org/browse/JDK-8295267): Improve G1 scan to merge ratio calculation

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [3ff69963](https://git.openjdk.org/jdk/pull/10694/files/3ff6996304b192553a8e429ea3e0b29682368eae)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [b648df4a](https://git.openjdk.org/jdk/pull/10694/files/b648df4a81e1dac0a2befb76ab50564507965658)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10694/head:pull/10694` \
`$ git checkout pull/10694`

Update a local copy of the PR: \
`$ git checkout pull/10694` \
`$ git pull https://git.openjdk.org/jdk pull/10694/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10694`

View PR using the GUI difftool: \
`$ git pr show -t 10694`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10694.diff">https://git.openjdk.org/jdk/pull/10694.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1Policy.cpp [hunk 1]
```diff
@@ -833,14 +835,12 @@ void G1Policy::record_young_collection_end(bool concurrent_operation_is_full_mar
...
-    // Due to duplicates in the log buffers, the number of actually scanned cards
+    // Due to duplicates in the log buffers, the number of scanned cards
     // can be smaller than the cards in the log buffers.
-    const size_t from_rs_length_cards = (total_cards_scanned > total_log_buffer_cards) ? total_cards_scanned - total_log_buffer_cards : 0;
+    const size_t scanned_cards_from_rs = (total_cards_scanned > merged_cards_from_log_buffers) ? total_cards_scanned - merged_cards_from_log_buffers : 0;
     double merge_to_scan_ratio = 0.0;
-    if (total_cards_scanned > 0) {
-      merge_to_scan_ratio = (double) from_rs_length_cards / total_cards_scanned;
+    if (merged_cards_from_rs > 0) {
+      merge_to_scan_ratio = (double)scanned_cards_from_rs / merged_cards_from_rs;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1Policy.cpp:843]: The name is probably wrong; a to b ratio is a/b, I believe.

**Review by albertnetymk — APPROVED** [review]
Some renaming suggestions, very subjective.

```
merged_cards_from_log_buffers -> cards_in_log_buffers
merged_cards_from_rs          -> cards_in_remset
total_cards_merged            -> total_cards
total_cards_scanned           -> scanned_cards
```

**Comment by tschatzl** [comment]
> Some renaming suggestions, very subjective.
> 
> ```
> merged_cards_from_log_buffers -> cards_in_log_buffers
> merged_cards_from_rs          -> cards_in_remset
> total_cards_merged            -> total_cards
> total_cards_scanned           -> scanned_cards
> ```

I kept the current naming as I believe these are too generic. Maybe that method should be split up somehow.

**Comment by tschatzl** [comment]
Thanks @albertnetymk @walulyai for your reviews
/integrate

**Pull request closed without merging**